bfvmm/include/intrinsics/x86/intel/apic/x2apic.h

```diff
@@ -2060,36 +2134,183 @@ namespace x2apic
 /// mode. It is marked final because it is intended to interact
 /// directly with x2apic hardware.
 ///
-struct EXPORT_X2APIC x2apic_control final : public lapic_control
-{
+struct EXPORT_X2APIC x2apic_control final : public lapic_control {
     //
     // Register reads
     //
-    value_type read_id() noexcept override;
-    value_type read_version() noexcept override;
-    value_type read_tpr() noexcept override;
-    value_type read_ldr() noexcept override;
-    value_type read_svr() noexcept override;
-    value_type read_icr() noexcept override;
+    value_type read_id() noexcept override
+    { return msrs::ia32_x2apic_apicid::get(); }
+
+    value_type read_version() noexcept override
+    { return msrs::ia32_x2apic_version::get(); }
+
+    value_type read_tpr() noexcept override
+    { return msrs::ia32_x2apic_tpr::get(); }
+
+    value_type read_ldr() noexcept override
+    { return msrs::ia32_x2apic_ldr::get(); }
+
+    value_type read_svr() noexcept override
+    { return msrs::ia32_x2apic_sivr::get(); }
 
-    value_type read_isr(index idx) noexcept override;
-    value_type read_tmr(index idx) noexcept override;
-    value_type read_irr(index idx) noexcept override;
+    value_type read_icr() noexcept override
+    { return msrs::ia32_x2apic_icr::get(); }
 
-    value_type read_lvt(lvt_reg reg) noexcept override;
-    value_type read_count(count_reg reg) noexcept override;
-    value_type read_div_config() noexcept override;
+
+    value_type read_isr(index idx) noexcept override
+    {
+        switch (idx) {
```

Comment:
use something like
auto addr = msrsr::ia32_x2apic::isr0::addr | idx
msrs::get(addr);

for isr, tmr, and irr. The addresses are sequential which is why this will work